client-js/src/client/errors.js

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright 2018, TeamDev. All rights reserved.
+ *
+ * Redistribution and use in source and/or binary forms, with or without
+ * modification, must retain the above copyright notice and the following
+ * disclaimer.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+* An error which occurs when sending off a request to Spine server endpoint.
+*
+* @abstract
+*/
+export class SpineWebError extends Error {
+
+  /**
+   * @param {!string} message the human-readable error message
+   * @param {*} cause         the reason why this error occurred
+   */
+  constructor(message, cause) {
+    super(message);
+    this.name = this.constructor.name;
+    this.cause = cause;
+  }
+
+  getCause() {
+    return this.cause;
+  }
+}
+
+/**
+ * An error which occurs when sending off a request to Spine server endpoint fails due to the
+ * connection problems. Combines situations in which the response to the sent request is not received.
+ *
+ * It may be caused by an incorrect server address or lack of network connectivity.
+ *
+ * @extends SpineWebError
+ */
+export class ConnectionError extends SpineWebError {
+
+  /**
+   * @param {!Error} error the error caught from {@code fetch} invocation
+   */
+  constructor(error) {
+    super(error.message, error);
+  }
+}
+
+/**
+ * An abstract error indicating an invalid server behaviour.
+ *
+ * @extends SpineWebError
+ * @abstract
+ */
+export class ServerError extends SpineWebError {
+}
+
+/**
+ * An abstract error indicating an invalid client behaviour.
+ *
+ * @extends SpineWebError
+ * @abstract
+ */
+export class ClientError extends SpineWebError {
+}
+
+/**
+ * An error which occurs when sending off a request to Spine server endpoint results
+ * with a response with `5xx` status code.
+ *
+ * Indicates an unhandled exception was thrown upon the request processing.
+ *
+ * @extends ServerError
+ */
+export class InternalServerError extends ServerError {
```

Comment:
I don’t think we need a server error hierarchy. We should be good with a single `ServerError` with response cause.